### ISSUE TITLE
Update by Schedule of stock index with more than 1000 entities

### ISSUE THREAD

**Issue by flauschi** [magento/magento2#22818]
### Preconditions (*)

1. Magento Community Edition (CE) 2.3.1

### Steps to reproduce (*)

1. Set mode of stock index (cataloginventory_stock) to "UPDATE BY SCHEDULE".
2. Have more than 1000 simple products waiting for reindex.
3. Wait for cron to finish reindex.

### Expected result (*)

1. The reindex should update the stock status for all products waiting for reindex.

### Actual result (*)

1. The reindex only updates stock status of the last products in the set (number of products % 1000)
2. The reindex deletes the stock status for all other products.
3. All the other products disappear on the frontend. They are also missing in grouped products.

### Possible fix
The problem is in [CatalogInventory/Model/ResourceModel/Indexer/Stock/DefaultStock.php](https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/CatalogInventory/Model/ResourceModel/Indexer/Stock/DefaultStock.php) in the method _updateIndex().

The call of _updateIndexTable() in the while loop inserts and updates data in portions of 1000 rows into the database. After the while loop the call of deleteOldRecords() deletes all these inserted und updated rows and _updateIndexTable() only inserts the rest of the data.

I think deleteOldRecords should only delete these records which were not updated by _updateIndexTable() (e.g. deleted products):
```
    protected function _updateIndex($entityIds)
    {
        $connection = $this->getConnection();
        $select = $this->_getStockStatusSelect($entityIds, true);
        $select = $this->getQueryProcessorComposite()->processQuery($select, $entityIds, true);
        $query = $connection->query($select);

        $i = 0;
        $data = [];
        $entityIdsTmp = [];
		
        while ($row = $query->fetch(\PDO::FETCH_ASSOC)) {
            $i++;
            $data[] = [
                'product_id' => (int)$row['entity_id'],
                'website_id' => (int)$row['website_id'],
                'stock_id' => Stock::DEFAULT_STOCK_ID,
                'qty' => (double)$row['qty'],
                'stock_status' => (int)$row['status'],
            ];
            $entityIdsTmp[] = (int)$row['entity_id'];
			
            if ($i % 1000 == 0) {
                $this->_updateIndexTable($data);
                $data = [];
            }
        }

        $this->deleteOldRecords(array_diff($entityIds, $entityIdsTmp));
        $this->_updateIndexTable($data);

        return $this;
    }
```

**Comment by engcom-Alfa** [comment]
Hi @flauschi ! Thank you for your report.

I am not able to reproduce the issue on fresh 2.3-develop and 2.3.2 instances.

This issue was already fixed in 2.3.2 release in commit 5ec74e261a8ff78a7c8a9b85b9344418df66ead3

**Comment by flauschi** [comment]
It was not fixed in 2.3.2. The commit 5ec74e2 was after 2.3.2 release. But it is true, that it is fixed in the develop branch. Thanks for closing the issue!
